src/Compilers/CSharp/Portable/Symbols/TypeParameterSymbol.cs

```diff
@@ -454,9 +444,43 @@ internal bool IsReferenceTypeFromConstraintTypes(ImmutableArray<TypeSymbolWithAn
             return AnyConstraintTypes(constraintTypes, inProgress, (type, arg) => ConstraintImpliesReferenceType(type.TypeSymbol, arg));
         }
 
-        internal bool IsNotNullableIfReferenceTypeFromConstraintTypes(ImmutableArray<TypeSymbolWithAnnotations> constraintTypes, ConsList<TypeParameterSymbol> inProgress)
+        internal bool? IsNotNullableIfReferenceTypeFromConstraintTypes(ImmutableArray<TypeSymbolWithAnnotations> constraintTypes, ConsList<TypeParameterSymbol> inProgress)
         {
-            return AnyConstraintTypes(constraintTypes, inProgress, (type, arg) => ConstraintImpliesIsNotNullableIfReferenceType(type, arg));
+            Debug.Assert(!constraintTypes.IsDefaultOrEmpty);
+            inProgress = inProgress.Prepend(this);
+
+            bool? result = false;
+            foreach (TypeSymbolWithAnnotations constraintType in constraintTypes)
+            {
+                if (constraintType.IsAnnotated)
+                {
+                    continue;
+                }
+
+                if (constraintType.TypeKind == TypeKind.TypeParameter)
+                {
+                    bool? isNotNullableIfReferenceType = ((TypeParameterSymbol)constraintType.TypeSymbol).GetIsNotNullableIfReferenceType(inProgress);
+
+                    if (isNotNullableIfReferenceType == false)
+                    {
+                        continue;
+                    }
+                    else if (isNotNullableIfReferenceType == null)
+                    {
+                        result = null;
+                        continue;
+                    }
+                }
+
+                if (constraintType.NonNullTypesContext.NonNullTypes == true)
+                {
+                    return true;
+                }
+
+                result = null;
```

Comment:
>result = null; [](start = 16, length = 14)

nit: this behavior should be documented: if any oblivious type constraint, then ...

I'm still not sure how to get two type constraints where one is oblivious and the other is not, but (if that's possible) I would think that the more constraining one (ie. the one that isn't oblivious) should win. But here we return `null` on the first oblivious type constraint.

Reply:
> but (if that's possible) I would think that the more constraining one (ie. the one that isn't oblivious) should win. 

Type parameters are different. With type parameters, the '?' doesn't constrain anything. The lack of '?' is more constraining.

---
In reply to: [213858018](https://github.com/dotnet/roslyn/pull/29586#discussion_r213858018) [](ancestors = 213858018)